argo/src/Argo.hs

```diff
@@ -339,6 +361,83 @@ instance JSON.ToJSON Request where
       "id"      .= view requestID     req <>
       "params"  .= view requestParams req
 
+data Response a
+  = Response { _responseID :: RequestID
+             , _responseAnswer :: a
+             , _responseStdOut :: String
+             , _responseStdErr :: String
+             } deriving (Eq, Ord, Show)
+
+responseID :: Simple Lens (Response a) RequestID
+responseID = lens _responseID (\r m -> r { _responseID = m })
+
+responseAnswer :: Simple Lens (Response a) a
+responseAnswer = lens _responseAnswer (\r m -> r { _responseAnswer = m })
+
+responseStdOut :: Simple Lens (Response a) String
+responseStdOut = lens _responseStdOut (\r m -> r { _responseStdOut = m })
+
+responseStdErr :: Simple Lens (Response a) String
+responseStdErr = lens _responseStdOut (\r m -> r { _responseStdOut = m })
+
+instance JSON.FromJSON a => JSON.FromJSON (Response a) where
+  parseJSON =
+    JSON.withObject ("JSON-RPC" <> T.unpack jsonRPCVersion <> " request") $
+    \o ->
+      (o .: "jsonrpc" `suchThat` (== jsonRPCVersion)) *>
+      (do result <- o .: "result"
+          Response
+            <$> o .: "id"
+            <*> result .: "answer"
+            <*> result .: "stdout"
+            <*> result .: "stderr")
+
+instance JSON.ToJSON a => JSON.ToJSON (Response a) where
+  toJSON resp =
+    JSON.object
+      [ "jsonrpc" .= jsonRPCVersion
+      , "id"      .= view responseID resp
+      , "result"  .= JSON.object
+        [ "answer"  .= view responseAnswer resp
+        , "stdout"  .= view responseStdOut resp
+        , "stderr"  .= view responseStdErr resp ] ]
+
+runMethodResponse ::
+  Method s a -> RequestID -> (Text -> IO ()) -> s -> IO (s, Response a)
+execMethodSilently ::
+  Method s a -> (Text -> IO ()) -> s -> IO s
+(runMethodResponse, execMethodSilently) =
+  ( \m reqID logger st ->
+      tryOutErr hCapture reqID (runMethod m logger st)
+  , \m logger st -> fst <$>
+      tryOutErr
+        (\hs a -> ("",) <$> hSilence hs a)
+        IDNull
+        (runMethod m logger st) )
+  where
+    tryOutErr ::
+      (forall a. [Handle] -> IO a -> IO (String, a)) ->
+      RequestID ->
+      IO (s, a) ->
+      IO (s, Response a)
+    tryOutErr capturer reqID action =
+      do (err, (out, res)) <-
+           capturer [stderr] . capturer [stdout] $
+           try @SomeException (try @JSONRPCException action)
+         tryPutOutErr out err
+         either
```

Comment:
I suspect this might be easier to follow as a `case` expression, but I don't feel strongly about it.